### ISSUE TITLE
Nomad interpolation not supported when using templating 

### ISSUE THREAD

**Issue by paladin-devops** [hashicorp/waypoint#1615]
I've been attempting to adopt the use of the nomad-jobspec "deployer" with the release of 0.4.0. My team and I thought this would be great for managing our Nomad deployments by deploying our existing Nomad job files. Variables are supplied to the existing Nomad job files; however when using templating in conjunction with nomad-jobspec, the variables are not being interpolated properly. Namely in the output below, the Nomad metadata interpolation is being mistaken as Waypoint variable interpolation.

Is there a way to support templating Nomad jobs with Waypoint input variables (or even allowing a Nomad input variable file?) AND simultaneously support Nomad variable interpolation?

Relevant documentation:
- https://www.waypointproject.io/plugins/nomad#nomad-jobspec-platform
- https://www.waypointproject.io/docs/waypoint-hcl/functions/template#templatestring
- https://www.nomadproject.io/docs/runtime/interpolation

waypoint.hcl contents:
```
project = "prometheus"

app "prometheus" {
  build {
    use "docker-pull" {
      image = "docker.io/prom/prometheus"
      tag   = "latest"
      encoded_auth = filebase64("${path.app}/auth.json")
    }
  }

  deploy {
    use "nomad-jobspec" {
      jobspec = templatestring(file("${path.app}/prometheus.nomad.hcl"), { //line 22
      job_env = "staging"
      job_name = "prometheus"
      job_dcs = "dc1"
      job_type = "service"
      count    = 1
      })
    }
  }
}
```

prometheus.nomad.hcl snippet

```
//
job "prometheus-server" {
  name = "${job_name}"
  datacenters = ["${job_dcs}"]
  type        = "${job_type}"

  constraint {
    attribute = "${meta.my-meta}" //line 51
    value = "on"
  }
//
```

"waypoint up" output:

```
» Deploying...
! /home/devops/waypoint/poc/waypoint.hcl:22,74-75:
  Invalid function argument; Invalid value for "vars" parameter: vars map does not
  contain key "meta", referenced at template:51,20-24., and 1 other diagnostic(s)
```

* Waypoint CLI Version: 0.4.0
* Waypoint Server Platform: Nomad
* Waypoint Plugin: Nomad

**Comment by mitchellh** [comment]
Ah yes, we're running into Nomad and Waypoint both using HCL. I think you have to escape your Nomad-specific interpolations, so change `${meta.` to `$${meta.`. This is because Waypoint has to template its stuff in first, followed by Nomad's. A bit ugly, but we're nesting the same templating formats here.

**Comment by paladin-devops** [comment]
That worked! Thanks a lot Mitchell. First app deployed w/Waypoint

**Comment by paladin-devops** [comment]
@mitchellh I've got a more interesting interpolation situation for ya this time. In this case, I'm using templatefile around my Nomad jobspec file, and I'm passing in some vars to the jobspec that way. However, while doing this, I can no longer use "${artifact.image}:${artifact.tag}" to specify my image in the jobspec, as it results in:

`Invalid function argument; Invalid value for "vars" parameter: vars map does not contain key "artifact"`

Here's what the line in the jobspec looks like:
`task "php" {
      driver = "docker"
      config {
        image = "${artifact.image}:${artifact.tag}"`

And in waypoint.hcl:
`  deploy {
    use "nomad-jobspec" {
      jobspec = templatefile("${path.app}/my-job.waypoint.nomad.hcl", {
        job_env         = workspace.name
        docker_password = var.docker_password #to pass into the auth stanza of Nomad's Docker driver config
        nomad_region    = var.nomad_region
      })
    }
  }`
  
  Using $$ in this case makes Nomad treat the image/tag as literally "${artifact.image}:${artifact.tag}", which I kind of assumed would happen but thought I'd at least give it a shot first. 
  
  Another possibly relevant detail is that I'm using the docker-pull build plugin. I did not experience this issue previously while using the regular docker build plugin though. I'm doing it this way now as a work around to the absence of a promotion feature by using docker-pull in my higher envs, while having development build the actual initial image.

**Comment by mitchellh** [comment]
I think the `templatefile` only has access to variables defined in the map there. It has access to the _functions_ available in HCL but not all the normal variables. So you'd have to pass them in `artifact = artifact` in your map there I think.

**Comment by paladin-devops** [comment]
@mitchellh finally had time to play around with this today and got it working. ```artifact = artifact``` being passed didn't solve the issue; but it did lead me to the root cause, which was that the "artifact" variable itself was not being defined/passed to the deploy stanza, because I did not have a "registry" stanza. As I mentioned in my last comment, in my other waypoint.hcl with the "regular" docker build plugin, I'm able to freely reference the "artifact" variable in my Nomad jobspec without passing in artifact in the map to templatefile. After I added in the registry stanza to my docker-pull Waypoint build, the artifact variable worked just fine.

While this works, it seems redundant to run a docker-pull, only to then push it right back up to the same registry from which I just pulled, so that my deployment can use it. This is probably asking for a lot under the hood, but maybe the build stanza by itself could set the "artifact" variable without a registry stanza? Nonetheless, working now!
